From 597b9b540c8043550437ca0600073b8579bc37c8 Mon Sep 17 00:00:00 2001
From: Alexandru Carp <alexandrucarp02@gmail.com>
Date: Wed, 12 Mar 2025 18:16:39 +0200
Subject: [PATCH 05/27] register analyzer process via netlink message, map
 intercepted calls into it, wait for response. also added netlink defines to a
 separate file so it can be used by both lsm and user-space program

---
 include/uapi/linux/mcheck.h |  20 +++++
 security/mcheck/mcheck.c    | 163 +++++++++++++++++++++++++++++-------
 2 files changed, 154 insertions(+), 29 deletions(-)
 create mode 100644 include/uapi/linux/mcheck.h

diff --git a/include/uapi/linux/mcheck.h b/include/uapi/linux/mcheck.h
new file mode 100644
index 000000000..095a71510
--- /dev/null
+++ b/include/uapi/linux/mcheck.h
@@ -0,0 +1,20 @@
+#define GEN_NETLINK_FAMILY_NAME "lsm_netlink"
+#define GEN_NETLINK_GROUP_NAME "lsm_mc_group"
+
+// Attributes (what we send/receive)
+enum {
+    LSM_ATTR_UNSPEC,
+    LSM_ATTR_ADDRESS,  // Memory address (unsigned long)
+    LSM_ATTR_RESPONSE, // User-space response (unsigned long)
+    __LSM_ATTR_MAX,
+};
+#define LSM_ATTR_MAX (__LSM_ATTR_MAX - 1)
+
+// Commands (message types)
+enum {
+    LSM_CMD_UNSPEC,
+    LSM_CMD_NOTIFY,  // Kernel → User: Send memory address
+    LSM_CMD_REPLY,   // User → Kernel: Receive analyzer response
+    __LSM_CMD_MAX,
+};
+#define LSM_CMD_MAX (__LSM_CMD_MAX - 1)
\ No newline at end of file
diff --git a/security/mcheck/mcheck.c b/security/mcheck/mcheck.c
index c557ab9de..cd1e0b998 100644
--- a/security/mcheck/mcheck.c
+++ b/security/mcheck/mcheck.c
@@ -1,43 +1,33 @@
+#include "linux/completion.h"
+#include "linux/kthread.h"
+#include "linux/mm.h"
+#include "linux/mm_types.h"
+#include "linux/mman.h"
 #include "linux/printk.h"
-#include "net/netlink.h"
+#include "linux/sched.h"
 #include <linux/security.h>
 #include <linux/export.h>
 #include <linux/lsm_hooks.h>
 #include <uapi/linux/lsm.h>
 #include <linux/netlink.h>
 #include <net/genetlink.h>
+#include <linux/mcheck.h>
 
 #ifdef pr_fmt
 #   undef pr_fmt
 #endif
 #define pr_fmt(fmt) KBUILD_MODNAME ": " fmt "\n"
 
-#define GEN_NETLINK_FAMILY_NAME "lsm_netlink"
+static struct task_struct *analyzer_task = NULL;
+static unsigned int analyzer_response = 0;
+static struct completion analyzer_ready;
 
-// Attributes (what we send/receive)
-enum {
-    LSM_ATTR_UNSPEC,
-    LSM_ATTR_ADDRESS,  // Memory address (unsigned long)
-    LSM_ATTR_RESPONSE, // User-space response (int)
-    __LSM_ATTR_MAX,
-};
-#define LSM_ATTR_MAX (__LSM_ATTR_MAX - 1)
-
-// Commands (message types)
-enum {
-    LSM_CMD_UNSPEC,
-    LSM_CMD_NOTIFY,  // Kernel → User: Send memory address
-    LSM_CMD_REPLY,   // User → Kernel: Receive integer response
-    __LSM_CMD_MAX,
-};
-#define LSM_CMD_MAX (__LSM_CMD_MAX - 1)
 
 // Attribute policy (validation)
 static struct nla_policy lsm_policy[LSM_ATTR_MAX + 1] = {
-    [LSM_ATTR_ADDRESS] = { .type = NLA_U32 },
-    [LSM_ATTR_RESPONSE] = { .type = NLA_S32 },
+    [LSM_ATTR_ADDRESS] = { .type = NLA_U64 },
+    [LSM_ATTR_RESPONSE] = { .type = NLA_U32 },
 };
-
 static int lsm_receive_reply(struct sk_buff *skb, struct genl_info *info);
 // Command handler mapping
 static const struct genl_ops lsm_ops[] = {
@@ -50,7 +40,7 @@ static const struct genl_ops lsm_ops[] = {
 
 static const struct genl_multicast_group genl_lsm_mcgrps[] = {
     {
-        .name = "lsm_mc_group"
+        .name = GEN_NETLINK_GROUP_NAME,
     },
 };
 
@@ -73,9 +63,24 @@ static int lsm_receive_reply(struct sk_buff *skb, struct genl_info *info)
         return -EINVAL;
     }
 
-    int user_response = nla_get_s32(info->attrs[LSM_ATTR_RESPONSE]);
+    int user_response = nla_get_u32(info->attrs[LSM_ATTR_RESPONSE]);
     pr_info("LSM: Received response from user-space: %d\n", user_response);
 
+    // register the analyzer task
+    if (!analyzer_task) {
+        pr_info("LSM: Received analyzer PID: %d\n", user_response);
+        analyzer_task = find_task_by_vpid(user_response);
+        if (!analyzer_task) {
+            pr_err("LSM: Failed to find analyzer task\n");
+            return -EINVAL;
+        }
+        return 0;
+    }
+
+    // set response and signal ready for lsm hook to continue
+    analyzer_response = user_response;
+    complete(&analyzer_ready);
+
     return 0;
 }
 
@@ -100,7 +105,7 @@ static int lsm_send_address(unsigned long addr) {
     }
 
     // Add the address attribute to the message
-    ret = nla_put_u32(skb, LSM_ATTR_ADDRESS, addr);
+    ret = nla_put_u64_64bit(skb, LSM_ATTR_ADDRESS, addr, 0);
     if (ret) {
         pr_err("LSM: Failed to add address attribute to Netlink message\n");
         genlmsg_cancel(skb, msg_head);
@@ -126,8 +131,8 @@ static int lsm_send_address(unsigned long addr) {
 
 static int mcheck_mmap_addr(unsigned long addr)
 {
-    pr_info("mcheck_mmap_addr: addr = %lx\n", addr);
-    lsm_send_address(addr);
+    // pr_info("mcheck_mmap_addr: addr = %lx\n", addr);
+    // lsm_send_address(addr);
     return 0;
 }
 static int mcheck_file_mprotect(struct vm_area_struct *vma, unsigned long reqprot, unsigned long prot)
@@ -136,10 +141,108 @@ static int mcheck_file_mprotect(struct vm_area_struct *vma, unsigned long reqpro
     return 0;
 }
 
+struct completion mmap_analyzer_thread_ready;
+struct mmap_analyzer_thread_data {
+    struct mm_struct *analyzer_mm;
+    unsigned long len;
+    struct vm_area_struct *vma;
+
+    unsigned long address;
+};
+static int mmap_analyzer_thread(void *arg) {
+    complete(&mmap_analyzer_thread_ready);
+
+    int ret = 0;
+    struct mmap_analyzer_thread_data *data = (struct mmap_analyzer_thread_data *)arg;
+
+
+    // kthread_use_mm can only be called from a kernel thread
+    kthread_use_mm(data->analyzer_mm);
+    data->address = vm_mmap(data->vma->vm_file, 0, data->len, PROT_READ, MAP_PRIVATE, 0);
+    if (!data->analyzer_mm) {
+        pr_err("mcheck_custom_mmap_hook: failed to map file\n");
+        ret = -ENOMEM;
+    }
+    pr_info("mapped address: %lx\n", data->address);
+    kthread_unuse_mm(data->analyzer_mm);
+    mmput(data->analyzer_mm);
+
+
+    while(!kthread_should_stop()) {
+        schedule();
+    }
+    return ret;
+}
+
+
 static int mcheck_custom_mmap_hook(unsigned long addr, unsigned long len, unsigned long prot)
 {
-    //pr_info("mcheck_custom_mmap_hook: addr = %lx, len = %lx, prot = %lx\n", addr, len, prot);
-    return 0;
+    // pr_info("mcheck_custom_mmap_hook: addr = %lx, len = %lx, prot = %lx\n", addr, len, prot);
+
+    // analyzer task is not set yet
+    if (!analyzer_task) {
+        return 0;
+    }
+
+    // we are only interested in executable mappings
+    if (!(prot & PROT_EXEC)) {
+        return 0;
+    }
+
+    // mapping in analyzer's address space
+    if (current->active_mm == analyzer_task->mm) {
+        return 0;
+    }
+
+    struct vm_area_struct *vma = find_vma(current->mm, addr);
+    if (!vma) {
+        pr_err("mcheck_custom_mmap_hook: failed to find vma\n");
+        return -EINVAL;
+    }
+
+    // we are only interested in file-backed mappings
+    if (!(vma->vm_file)) {
+        return 0;
+    }
+
+    struct mm_struct *analyzer_mm = get_task_mm(analyzer_task);
+    if (!analyzer_mm) {
+        pr_err("mcheck_custom_mmap_hook: failed to get analyzer mm\n");
+        return -EINVAL;
+    }
+
+    struct mmap_analyzer_thread_data *analyzer_data = kmalloc(sizeof(struct mmap_analyzer_thread_data), GFP_KERNEL);
+    if (!analyzer_data) {
+        pr_err("mcheck_custom_mmap_hook: failed to allocate memory\n");
+        return -ENOMEM;
+    }
+    analyzer_data->analyzer_mm = analyzer_mm;
+    analyzer_data->len = len;
+    analyzer_data->vma = vma;
+
+    init_completion(&mmap_analyzer_thread_ready);
+    struct task_struct *task = kthread_run(mmap_analyzer_thread, analyzer_data, "mmap_analyzer_thread");
+
+    if (IS_ERR(task)) {
+        pr_err("Failed to create kernel thread: %ld\n", PTR_ERR(task));
+        kfree(analyzer_data);
+        return 0;
+    }
+
+    wait_for_completion(&mmap_analyzer_thread_ready);
+
+    int ret = kthread_stop(task);
+
+    if (ret) {
+        pr_err("Failed to stop kernel thread: %d\n", ret);
+    } else {
+        lsm_send_address(analyzer_data->address);
+        wait_for_completion(&analyzer_ready);
+        reinit_completion(&analyzer_ready);
+    }
+
+    kfree(analyzer_data);
+    return -analyzer_response;
 }
 
 static int mcheck_custom_mprotect_hook(unsigned long addr, unsigned long len, unsigned long prot)
@@ -169,6 +272,8 @@ static int __init mcheck_lsm_init(void)
         pr_err("LSM: Failed to register Netlink family\n");
         return ret;
     }
+
+    init_completion(&analyzer_ready);
     pr_info("MCHECK: LSM initialized\n");
     return 0;
 }
-- 
2.54.0

